gramps/gen/filters/_genericfilter.py

```diff
@@ -282,9 +278,37 @@ def apply(self, db, id_list=None, tupleind=None, user=None, tree=False):
             raise Exception("invalid operator: %r" % self.logical_op)
 
         start_time = time.time()
-        res = self.apply_logical_op_to_all(
-            db, id_list, apply_logical_op, user, tupleind, tree
-        )
+
+        # build the starting set of possible_handles to be filtered
+        possible_handles: Set[PrimaryObjectHandle]
+        if id_list:
+            if tupleind:
+                # construct a dict from handle to corresponding tuple
+                # this is used to efficiently transform final_list from a list of
+                # handles to a list of tuples
+                handle_tuple: Dict[PrimaryObjectHandle, Tuple] = {
+                    data[tupleind]: data for data in cast(List[Tuple], id_list)
+                }
+                possible_handles = set(handle_tuple.keys())
+            else:
+                possible_handles = set(cast(List[PrimaryObjectHandle], id_list))
+        elif tree:
+            tree_handles = [handle for handle, obj in self.get_tree_cursor(db)]
+            possible_handles = set(tree_handles)
+        else:
+            possible_handles = set(self.get_all_handles(db))
```

Comment:
We already have a method specific to the filter type. Let's use it, and make it parallel with the tree case:

```suggestion
            all_handles = [handle for handle, obj in self.get_cursor(db)]
            possible_handles = set(all_handles)
```

Reply:
Speed was my concern. `get_all_handles` is O(1) which is better the O(n) of using a `cursor`. The `cursor` also constructs the object which we do not need at this stage.

Reply:
Unless you feel strongly, I will leave the new `get_all_handles` methods

Reply:
Yes, you have a point. Actually I think they are both O(n), but one is in SQL and the other is in Python. Let's leave it as you have it.